### DISCUSSION TITLE
Rolling positions with help of a FlashLoanProvider

### DISCUSSION THREAD

**Discussion by samclassix** [Frankencoin-ZCHF/Frankencoin#28] [Ideas]
# Development Branch: **samclassix-flash-loan-provider**

> This is just a "Proof of Concept" branch and further developments can/should be made before merging.
https://github.com/Frankencoin-ZCHF/FrankenCoin/tree/samclassix-flash-loan-provider/contracts/FlashLoanRoller

# Why is this wanted and needed?

The idea revolves around using the FlashLoan mechanism to enable users to roll over their positions in the Frankencoin ecosystem without requiring additional liquidity ~~or as little as the flash fee (0.1% used in test deployment)~~. Here's an expanded discussion of the concept, potential implementation, and implications for FPS holders and ZCHF holders.

> **hint:** When we talk about "rolling" we straight away limit "us" in thinking what is possible with this mechanism. Just to make you aware of what I mean, lets replace the wording "rolling a position" with "merging a position". So basically the mechanic is **merge** and not really **roll**. So you could also merge positions with the same collateral, if collateral balance fits all mint ratios. :)  

# Why a Flash Loan?

> Position A --> Position B 
> From (ref. existing/old) --> To (ref. new)

In essence, users could 

- borrow ZCHF tokens via a flash loan 
- to repay an existing and outstanding mint, 
- then immediately send the collateral from the "old" position to the "new" position, 
- mint as much as needed
   - incl. Interest
   - incl. Reserve
   - incl. pay out amount
- repaying the loan
- paying the flash fee
within the same transaction. (atomic flash loan)
 
This process would be particularly beneficial for users needing to adjust position parameters or extend the duration of their position without the friction of closing and reopening.

# Who could potentially benefit from a Flash Loan (two approaches)?

1. FPS holders, who would earn the fee
2. ZCHF holders, who could deposit zchf to a pool for flash loans and receive the fees

There are pro and cons with both approaches. The first approach could be simpler in development due to mint and burn mechanisms and with "minters" power claiming ZCHF without needing to approve. We will go with this one for now and since its development is almost done. Yesterday i managed to roll positions with the ability of flash loans on a development deployment.

# Impact for FPS holders

FPS holder would benefit, due to
- control to veto the FlashLoanProvider when it gets proposed as a minter
- control over shutdown with veto power, after proposal

- control over approving limit via the proposal, e.g.
> FLASHLOAN_TOTALMAX: 10_000_000 ZCHF
> FLASHLOAN_MAX: 1_000_000 ZCHF
> FLASHLOAN_FEE:  0.1 %

- earn a fee, e.g. 0.1% which gets transferred directly to the reserve

# Quick overview of the effect of rolling a position 

(simplified)
- reserve 10%
- 1 coin @ 10000 zchf

**Owner to Roller**
additionalCollateral (make any transfers before): e.g. 0 coin
flashFee (allowance): e.g. 9 zchf
> 0.1% of (minted - reserve) = 0.001 * (10000 - 1000) = 9

**From Position**
minted: e.g. 10000 zchf --> 0 zchf 
collateral: e.g. 100 coin --> 0 coin

**To Position**
minted. e.g. 10000 zchf + interest --> eg. 10025,6321996571 zchf
collateral: e.g. 1 coin (e.g. min. collateral for created position) --> 101 coin

If the "to" position has already a mint, e.g. for merged roll into one position,
could be useful for:
- merging the from position into a other position
- "closing" or lets say "merge" them into the original position or different position

# How does the process work? 

1. Lets assume the frankencoin ecosystem is already fully deployed
2. FlashLoanProvider passed as minter, cooldown period passed. State: ready
3. User: creates **Position A**
> minCollateral: 1 coin
> sizeCollateral: 10 coin
> liqPrice: 6000 zchf / coin
> <ins>minted: 60000 zchf</ins>

4. User: decides to roll/merge into a new created position
5. User: creates **Position B** 
> minCollateral: 1 coin
> sizeCollateral: 1 coin
> liqPrice: 6000 zchf / coin
> <ins>minted: 0 zchf</ins>

7. Waits until the position is passed without veto (considered open)
8. User: creates a ownable **Roller** (belongs to an owner)
9. User: transferOwnership of **Position A** to the owned roller
10. User: transferOwnership of **Position B** to the owned roller 
11. User: roller.prepareAndExecute(address(**Position A**), address(**Position B**));

### After effect
- the roller will, collect all costs (flashFee, interestFee of new mint)
- the roller will, cover all costs during the process by minting "more" zchf
- the roller will, transferOwnership of both positions back to the user
> **hint:** you can always manually **redeemOwnership** of Ownable contracts

**Position A**
> minCollateral: 1 coin
> sizeCollateral: 0 coin **(will be considered as closed)**
> liqPrice: 6000 zchf / coin
> <ins>minted: 0 zchf</ins>

**Position B**
> minCollateral: 1 coin
> sizeCollateral: 11 coin
> liqPrice: 6000 zchf / coin
> <ins>minted: 60,213.9469911407 zchf</ins>

**Comment by samclassix** [comment]
# Example with merging positions
Meaning the target positions minted amount is larger then 0 zchf. (already minted some zchf).

> **YOU NEED "SOME" FREE MARGIN**, its not possible to merge two fully minted positions because noone would cover the flashFee or the additional interest by the positions mint. --> I will also test an example to merge two fully minted positions, by manually adding some collateral funds to the roller contract, before executing the merge.

## Before

**Position A**
minCollateral: 1 coin
sizeCollateral: 100 coin
liqPrice: 6000 zchf / coin
<ins>minted: 600000 zch</ins>

**Position B**
minCollateral: 1 coin
sizeCollateral: 1 coin
liqPrice: 6000 zchf / coin
<ins>minted: 1000 zchf</ins> <--- <ins>See, there is a margin left to mint. That is needed.</ins>

## After

**Position A**
minCollateral: 1 coin
<ins>sizeCollateral: 0 coin</ins>
liqPrice: 6000 zchf / coin
<ins>minted: 0 zch</ins>

**Position B**
minCollateral: 1 coin
<ins>sizeCollateral: 101 coin</ins>
liqPrice: 6000 zchf / coin
<ins>minted: 603139.4699 zchf</ins>

**Flash Loan**
<ins>flashLoan: 540000 zchf</ins> (hint: reduced by the reserve amount - 10%)
<ins>flashFee: 540 zchf</ins> (0.1% fee)

```
"event": "MintingUpdate",
"args": {
	"collateral": "101000000000000000000",
	"price": "6000000000000000000000",
	"minted": "603139469911406830129030"
}

"event": "Repaid",
"args": {
	"from": "0x6e44CEbcf2F1BB4eb7e59Bf1171757c91A803900",
	"total": "540540000000000000000000",
	"repay": "540000000000000000000000",
	"fee": "540000000000000000000"
}
```

**Comment by samclassix** [comment]
# Example with two fully minted position

> this is only possible, if you add enough collateral to the roller contract, before executing

```
**Error provided by the contract: InsufficientCollateral**

If you do not add additional collateral, the FlashLoanRoller & FlashLoanProvider will still pass, however, the error will be triggered in the logic of the position itself (what is wanted).

The position won't allow to merge, if there is not enough collateral to cover the flashFee and the additional minting fee of the position.
```

## Before

**Position A**
minCollateral: 1 coin
<ins>sizeCollateral: 100 coin</ins>
liqPrice: 6000 zchf / coin
<ins>minted: 600000 zch</ins>

**Position B**
minCollateral: 1 coin
sizeCollateral: 100 coin
liqPrice: 6000 zchf / coin
<ins>minted: 600000 zchf</ins> <--- <ins> There is NOT margin left to mint.</ins>

**Additional Collateral**
addCollateral: 2 coin (should be enough)

## After

**Position A**
minCollateral: 1 coin
<ins>sizeCollateral: 0 coin</ins>
liqPrice: 6000 zchf / coin
<ins>minted: 0 zch</ins>

**Position B**
minCollateral: 1 coin
sizeCollateral: 202 coin
liqPrice: 6000 zchf / coin
<ins>minted: zchf</ins> <--- <ins>

## Test workflow in solidity

```
function initB_Positions() public {
    posFrom = Position(addPosition(100 ether)); // create position with 100 coins
    posTo = Position(addPosition(100 ether)); // create position with 100 coins
}

function initC_Minting() public {
    posFrom.adjust(600000 ether, coin.balanceOf(address(posFrom)), posFrom.price()); // max mint
    posTo.adjust(600000 ether, coin.balanceOf(address(posTo)), posTo.price()); // max mint
}

function initD_Flashloan() public {
    coin.transfer(address(roller), 2 ether); // add 2 additional coins

    posFrom.transferOwnership(address(roller)); // transfer ownership
    posTo.transferOwnership(address(roller)); // transfer ownership

    roller.prepareAndExecute(address(posFrom), address(posTo)); // execute merge of two fully minted positions
}
```

## Some relevant logs

#### Loan taken in zchf
```
"event": "LoanTaken",
"args": {
	"to": "0xA7c0D580D5e82f1910b4F31B40796431DBA77A76",
	"amount": "540000000000000000000000",
	"totalMint": "540000000000000000000000" <--- here
}
```

#### Repayment of zchf in "from" position
```
"event": "MintingUpdate",
"args": {
	"collateral": "100000000000000000000",
	"price": "6000000000000000000000",
	"minted": "0" <--- here
}
```

#### Withdrawal of collateral in "from" position
```
"event": "MintingUpdate",
"args": {
	"collateral": "0", <--- here
	"price": "6000000000000000000000",
	"minted": "0"
}
```

**Comment by luziusmeisser** [comment]
This has been implemented, although it is not exposed in the frontend yet.
